File: src/Algoritms/Genetic Algorithm/crossover.js
// ================================
// 1. Crossover for Binary Numbers
// ================================


// One-point crossover for binary representations

export function onePointCrossover(parent1, parent2) {
    const point = Math.floor(Math.random() * (parent1.length - 1)) + 1;
    const child1 = [...parent1.slice(0, point), ...parent2.slice(point)];
    const child2 = [...parent2.slice(0, point), ...parent1.slice(point)];
    return {
        offspring: [child1, child2],
        point: point
    };
}

// Two-point crossover for binary representations
 
export function twoPointCrossover(parent1, parent2) {
    let point1 = Math.floor(Math.random() * parent1.length);
    let point2 = Math.floor(Math.random() * (parent1.length - 1));
    if (point2 >= point1) point2++;
    else [point1, point2] = [point2, point1];
    
    const child1 = [
        ...parent1.slice(0, point1),
        ...parent2.slice(point1, point2),
        ...parent1.slice(point2)
    ];
    const child2 = [
        ...parent2.slice(0, point1),
        ...parent1.slice(point1, point2),
        ...parent2.slice(point2)
    ];
    return {
        offspring: [child1, child2],
        points: [point1, point2]
    };
}

// Uniform crossover for binary representations
 
export function uniformCrossover(parent1, parent2, crossoverRate = 0.5) {
    const child1 = [];
    const child2 = [];
    const swaps = [];
    
    for (let i = 0; i < parent1.length; i++) {
        if (Math.random() < crossoverRate) {
            child1.push(parent2[i]);
            child2.push(parent1[i]);
            swaps.push(i);
        } else {
            child1.push(parent1[i]);
            child2.push(parent2[i]);
        }
    }
    return {
        offspring: [child1, child2],
        swaps: swaps
    };
}


// ==============================
// 2. Crossover for Real Numbers
// ==============================


export function simpleCrossover(parent1, parent2, alpha = 0.5) { 
    const length = parent1.length;
    const point = Math.floor(Math.random() * (length - 1)) + 1; 

    const child1 = [...parent1.slice(0, point)];
    const child2 = [...parent2.slice(0, point)];
    
    // Blend remaining genes using alpha
    for (let i = point; i < length; i++) {
        child1.push(alpha * parent1[i] + (1 - alpha) * parent2[i]);
        child2.push(alpha * parent2[i] + (1 - alpha) * parent1[i]);
    }
    
    return {
        offspring: [child1, child2],
    };
}


export function simpleArithmeticCrossover(parent1, parent2, alpha = 0.5) {
    const length = parent1.length;
    const point = Math.floor(Math.random() * length); 
    
    const child1 = [...parent1];
    const child2 = [...parent2];

    child1[point] = alpha * parent1[point] + (1 - alpha) * parent2[point];
    child2[point] = alpha * parent2[point] + (1 - alpha) * parent1[point];

    return {
        offspring: [child1, child2],
    };
}


export function wholeArithmeticCrossover(parent1, parent2, { alpha = 0.5 } = {}) {
    const length = parent1.length;
    
    const child1 = [];
    const child2 = [];

    for (let i = 0; i < length; i++) {
        child1.push(alpha * parent1[i] + (1 - alpha) * parent2[i]);
        child2.push(alpha * parent2[i] + (1 - alpha) * parent1[i]);
    }

    return {
        offspring: [child1, child2],
    };
}


// ==============================
// 2. Crossover for Permutation
// ==============================
export function orderCrossover(parent1, parent2) {
    const length = parent1.length;
    
    let point1 = 3;
    let point2 = 6;

    let child1 = new Array(length).fill(-1);
    let child2 = new Array(length).fill(-1);

    for (let i = point1; i <= point2; i++) {
        child1[i] = parent1[i];
        child2[i] = parent2[i];
    }

    function fillRemaining(child, parent) {
        let index = (point2 + 1) % length; 
        for (let i = 0; i < length; i++) {
            let gene = parent[(point2 + 1 + i) % length];
            if (!child.includes(gene)) {
                child[index] = gene;
                index = (index + 1) % length; 
            }
        }
    }

    fillRemaining(child1, parent2);
    fillRemaining(child2, parent1);

    return {
        offspring: [child1, child2],
        crossoverPoints: [point1, point2]
    };
}

File: crossover.py
import random
from typing import List, Tuple, Union

# ================================
# 1. Crossover for Binary Numbers
# ================================

def one_point_crossover(parent1: List[int], parent2: List[int]) -> Tuple[List[int], List[int]]:
    """
    One-point crossover for binary representations.
    A single crossover point is selected, and tails are swapped.
    """
    point = random.randint(1, len(parent1) - 1)
    child1 = parent1[:point] + parent2[point:]
    child2 = parent2[:point] + parent1[point:]
    return child1, child2

def two_point_crossover(parent1: List[int], parent2: List[int]) -> Tuple[List[int], List[int]]:
    """
    Two-point crossover for binary representations.
    Two crossover points are selected, and the middle segment is swapped.
    """
    point1, point2 = sorted(random.sample(range(1, len(parent1)), 2))
    child1 = parent1[:point1] + parent2[point1:point2] + parent1[point2:]
    child2 = parent2[:point1] + parent1[point1:point2] + parent2[point2:]
    return child1, child2

def uniform_crossover(parent1: List[int], parent2: List[int], crossover_rate: float = 0.5) -> Tuple[List[int], List[int]]:
    """
    Uniform crossover for binary representations.
    Each gene is swapped with a probability of crossover_rate.
    """
    child1, child2 = [], []
    for i in range(len(parent1)):
        if random.random() < crossover_rate:
            child1.append(parent2[i])
            child2.append(parent1[i])
        else:
            child1.append(parent1[i])
            child2.append(parent2[i])
    return child1, child2

# ==============================
# 2. Crossover for Real Numbers
# ==============================

def simple_crossover(parent1: List[float], parent2: List[float]) -> Tuple[List[float], List[float]]:
    """
    Simple crossover for real numbers.
    A single crossover point is selected, and tails are swapped.
    """
    point = random.randint(1, len(parent1) - 1)
    child1 = parent1[:point] + parent2[point:]
    child2 = parent2[:point] + parent1[point:]
    return child1, child2

def simple_arithmetic_crossover(parent1: List[float], parent2: List[float], alpha: float = 0.5) -> Tuple[List[float], List[float]]:
    """
    Simple arithmetic crossover for real numbers.
    Each gene is a weighted average of the parents' genes.
    """
    child1 = [alpha * p1 + (1 - alpha) * p2 for p1, p2 in zip(parent1, parent2)]
    child2 = [alpha * p2 + (1 - alpha) * p1 for p1, p2 in zip(parent1, parent2)]
    return child1, child2

def whole_arithmetic_crossover(parent1: List[float], parent2: List[float], alpha: float = 0.5) -> Tuple[List[float], List[float]]:
    """
    Whole arithmetic crossover for real numbers.
    All genes are combined using a weighted average.
    """
    child1 = [alpha * p1 + (1 - alpha) * p2 for p1, p2 in zip(parent1, parent2)]
    child2 = [alpha * p2 + (1 - alpha) * p1 for p1, p2 in zip(parent1, parent2)]
    return child1, child2

# ============================
# 3. Crossover for Permutations
# =============================

def order_recombination(parent1: List[int], parent2: List[int]) -> Tuple[List[int], List[int]]:
    """
    Order recombination (OX) for permutation-based representations.
    A segment is copied from one parent, and the remaining genes are filled in order from the other parent.
    """
    size = len(parent1)
    point1, point2 = sorted(random.sample(range(size), 2))
    
    def create_child(p1, p2):
        child = [None] * size
        child[point1:point2] = p1[point1:point2]
        remaining = [gene for gene in p2 if gene not in child]
        idx = 0
        for i in range(size):
            if child[i] is None:
                child[i] = remaining[idx]
                idx += 1
        return child
    
    child1 = create_child(parent1, parent2)
    child2 = create_child(parent2, parent1)
    return child1, child2

def cycle_recombination(parent1: List[int], parent2: List[int]) -> Tuple[List[int], List[int]]:
    """
    Cycle recombination (CX) for permutation-based representations.
    Cycles of genes are alternated between parents.
    """
    size = len(parent1)
    child1, child2 = [None] * size, [None] * size
    idx = 0
    
    while None in child1:
        if child1[idx] is None:
            # Start a cycle
            current_idx = idx
            while child1[current_idx] is None:
                child1[current_idx] = parent1[current_idx]
                child2[current_idx] = parent2[current_idx]
                current_idx = parent1.index(parent2[current_idx])
        idx += 1
    
    return child1, child2

# ========================
# 4. Crossover Wrapper
# ========================

# def crossover(
#     parent1: Union[List[int], List[float]],
#     parent2: Union[List[int], List[float]],
#     method: str,
#     **kwargs
# ) -> Tuple[Union[List[int], List[float]], Union[List[int], List[float]]]:
#     """
#     Wrapper function to call the appropriate crossover method.
#     """
#     if method == "one_point":
#         return one_point_crossover(parent1, parent2)
#     elif method == "two_point":
#         return two_point_crossover(parent1, parent2)
#     elif method == "uniform":
#         return uniform_crossover(parent1, parent2, kwargs.get("crossover_rate", 0.5))
#     elif method == "simple":
#         return simple_crossover(parent1, parent2)
#     elif method == "simple_arithmetic":
#         return simple_arithmetic_crossover(parent1, parent2, kwargs.get("alpha", 0.5))
#     elif method == "whole_arithmetic":
#         return whole_arithmetic_crossover(parent1, parent2, kwargs.get("alpha", 0.5))
#     elif method == "order":
#         return order_recombination(parent1, parent2)
#     elif method == "cycle":
#         return cycle_recombination(parent1, parent2)
#     else:
#         raise ValueError(f"Unknown crossover method: {method}")